### DISCUSSION TITLE
Breaking up http wrappers into their own modules.

### DISCUSSION THREAD

**Discussion by elliotcourant** [getsentry/sentry-go#388] [Ideas]
Right now when you import `github.com/getsentry/sentry-go` you also get some of the dependencies for the HTTP libraries that sentry supports. For example;

https://github.com/getsentry/sentry-go/blob/bd116d6ce79b604297c6497aa07d7ac01768adbb/go.mod#L16

Any project that depends on `sentry-go` will now also depend on `kataras/iris`. Because of this; version of iris (or any other dependency that is imported by sentry-go to support http libraries) could include security vulnerabilities.

While it's unlikely that these vulnerabilities will cause direct problems; or even any problems at all. It still causes some oddities in GitHub's security notifications.

Right now there is CVE-2021-42576 which affects older versions kataras/iris. This was recently fixed in the latest commit https://github.com/kataras/iris/commit/485395190b77a8517ad0b6112eaea2d912cc066f but even if you upgrade your go project to use that commit. You will still have the security warning from GitHub because of the side affect imports from sentry-go

```bash
$ go mod graph | grep bluemonday
github.com/monetr/monetr github.com/microcosm-cc/bluemonday@v1.0.16
github.com/kataras/iris/v12@v12.2.0-alpha4.0.20211031094939-485395190b77 github.com/microcosm-cc/bluemonday@v1.0.16
github.com/kataras/iris/v12@v12.1.8 github.com/microcosm-cc/bluemonday@v1.0.2
github.com/kataras/iris/v12@v12.2.0-alpha2 github.com/microcosm-cc/bluemonday@v1.0.4

# ^ This shows that even though my package imports the latest version of the previously vulnerable package; there is still an a path in the dependency tree that requires an older version. An older version of iris is importing the older version of bluemonday

$ go mod graph | grep github.com/kataras/iris/v12@v12.1.8
github.com/getsentry/sentry-go@v0.11.0 github.com/kataras/iris/v12@v12.1.8

# ^ Looking further, the package that requires the older version of iris is sentry-go
```

So even if my project does _not_ import or use the iris package from within sentry-go. I still have a less than secure package in my dependency tree.

This is not a big issue with this particular package; but it does make me a bit worried about other packages that could be at risk in the future further down the dependency tree.

---

As a potential solution to this; the base `sentry-go` package does not need to depend on any of the HTTP libraries that it supports. Instead a go module could be created within the `sentry-go` repo for each of those HTTP libraries. With the dependencies for those libraries being specific to that module.

This way simply importing `sentry-go` does not give me the dependencies for `iris` or any other HTTP library unless I explicitly import them.

Curious on what Sentry's thoughts are on this?

**Comment by elliotcourant** [comment]
https://github.com/getsentry/sentry-go/pull/389 I've implemented a basic form of this in this PR. The same approach could be done for any of the http libraries that sentry-go supports.

On the current `master` branch:

```bash
$ go mod graph | grep bluemonday
github.com/kataras/iris/v12@v12.1.8 github.com/microcosm-cc/bluemonday@v1.0.2
github.com/microcosm-cc/bluemonday@v1.0.2 golang.org/x/net@v0.0.0-20181220203305-927f97764cc3
```

With sentryiris as a separate module:

```bash
$ go mod graph | grep bluemonday
# Nothing
```

**Comment by rhcarvalho** [comment]
Splitting into multiple modules is tracked by #156.

Note however that the work involved is unfortunately more than just adding `go.mod` files.
It requires changes to how we do releases (proper tagging for all modules, bumping `go.mod` versions for all "submodules"), how continuous integration in GitHub Actions is setup (must make sure to run tests for all modules, not only the main module), and even affects how development works (multi-module support in VScode is still not a great developer experience).

So this change has implications that go beyond a few files in the repo, and is difficult / impossible to undo if something goes wrong (proxies will cache broken releases). Unfortunately we don't have the bandwidth to do that right now properly.

You can see on the Internet how other projects struggle with multi-module setups, two recent links that come to mind:
- https://togithub.com/prometheus/prometheus/issues/8513
- https://changelog.com/gotime/188

> While it's unlikely that these vulnerabilities will cause direct problems; or even any problems at all. It still causes some oddities in GitHub's security notifications.

Go 1.17 finally landed [module graph pruning](https://golang.org/ref/mod#graph-pruning). So now more than ever GitHub's security notifications are just producing too many false positives as it treats [`go.sum` as a "lock file", something it is not](https://github.com/golang/go/wiki/Modules#is-gosum-a-lock-file-why-does-gosum-include-information-for-module-versions-i-am-no-longer-using).
